refactor: Improve special member functions definitions

hebasto opened the pull request:
This PR fixes special member functions in the following class/struct definitions:
- `struct PartiallySignedTransaction` (done in  #17349)
  The default (i.e., generated by a compiler) copy constructor does the same things. It prevents `-Wdeprecated-copy` warning in GCC 9 (#15822) and `-Wdeprecated` warning in Clang 10 for implicitly declared `operator=`
- `class SigningProvider`
  All special member functions of the SigningProvider class defined `=default` which makes it a well-behaved base class. It prevents `-Wdeprecated` warning in Clang 10
- `struct secure_allocator`
  Define `=default` copy assignment explicitly. It prevents `-Wdeprecated` warning in Clang 10
- `struct FrozenCleanupCheck`
  Define `=default` copy constructor explicitly. It prevents `-Wdeprecated` warning in Clang 10

emilengler commented:
Concept ACK 5b2bb0f
See 385-386 in src/psbt.h and 12-13 in src/psbt.cpp

practicalswift commented:
@hebasto Is this change complete in the sense that it addresses all instances of redundant copy constructors in our code base? :-)

hebasto commented:
@practicalswift 
> Is this change complete in the sense that it addresses all instances of redundant copy constructors in our code base? :-)

I don't know ((
What tool could be helpful to check our codebase? It seems GCC 9 compiler do this work, right?

practicalswift commented:
@hebasto Isn't it `-Wdeprecated-copy` in `gcc` and `-Wdeprecated` in `clang`? :-)

hebasto commented:
@practicalswift on master:
```
$ gcc --version | grep gcc
gcc (GCC) 9.1.1 20190503 (Red Hat 9.1.1-1)

$ make 2>&1 | grep -e "-Wdeprecated-copy" | grep -vE '(^/usr|leveldb)'
psbt.cpp:335:19: warning: implicitly-declared ‘PartiallySignedTransaction& PartiallySignedTransaction::operator=(const PartiallySignedTransaction&)’ is deprecated [-Wdeprecated-copy]
```

promag commented:
ACK 5b2bb0fd3af34463bb382b473cc8b6dee65c0920. Being complete is nice, disallowing more of these would be great.

fanquake commented:
I'm seeing more of these warnings when compiling using Apple Clang with `-Wdeprecated`:
```bash
./psbt.h:398:5: warning: definition of implicit copy assignment operator for 'PartiallySignedTransaction' is deprecated because it has a user-declared copy constructor [-Wdeprecated]
./script/signingprovider.h:21:13: warning: definition of implicit copy assignment operator for 'SigningProvider' is deprecated because it has a user-declared destructor [-Wdeprecated]
./script/signingprovider.h:21:13: warning: definition of implicit copy constructor for 'SigningProvider' is deprecated because it has a user-declared destructor [-Wdeprecated]
test/checkqueue_tests.cpp:115:5: warning: definition of implicit copy constructor for 'FrozenCleanupCheck' is deprecated because it has a user-declared destructor [-Wdeprecated]
./support/allocators/secure.h:35:5: warning: definition of implicit copy assignment operator for 'secure_allocator<unsigned char>' is deprecated because it has a user-declared destructor [-Wdeprecated]
```

```bash
clang -v
Apple LLVM version 10.0.1 (clang-1001.0.46.4)
Target: x86_64-apple-darwin18.7.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

hebasto committed:
refactor: Improve SigningProvider base class

All special member functions of the SigningProvider class defined 
=default which makes it a well-behaved base class.

hebasto committed:
refactor: Define =default copy assignment explicitly

hebasto committed:
refactor: Define =default copy ctor explicitly

hebasto commented:
@fanquake 
> I'm seeing more of these warnings when compiling using Apple Clang with `-Wdeprecated`...

With the latest push:
```sh
$ clang --version | grep version
Apple LLVM version 10.0.1 (clang-1001.0.46.4)
$ ./configure CXXFLAGS="-Wdeprecated"
...
$ make 2>&1 | grep -e "warning:" | grep -vE '^/usr|leveldb|univalue'
```
the output has no mentioned warnings.

hebasto commented:
The PR description has been updated.

hebasto commented:
@practicalswift Would you mind review this PR?

hebasto commented:
Rebased on top of #17349.

MarcoFalke commented on src/test/checkqueue_tests.cpp:
```diff
@@ -112,6 +112,7 @@ struct FrozenCleanupCheck {
         return true;
     }
     FrozenCleanupCheck() {}
```
why is this needed?

hebasto replied:
The default ctor is required at least here: https://github.com/bitcoin/bitcoin/blob/94a26b192f187cb50bf1ac1775b23f2b03f772b1/src/checkqueue.h#L150

practicalswift replied:
Unrelated, but is there any reason we're doing the extra move here (i.e. `push_back` instead of `emplace_back`)?

MarcoFalke replied:
> The default ctor is required at least here: 

Yeah, but how is the manually defined ctor different from the compiler generated?

hebasto replied:
https://en.cppreference.com/w/cpp/language/default_constructor

> If no user-declared constructors of any kind are provided for a class type (struct, class, or union), the compiler will always declare a default constructor as an inline public member of its class.

But we have a copy ctor with this PR, so the compiler will not generate a default ctor.

MarcoFalke replied:
So why do you provide the copy ctor? How is it different from the compiler generated one?

MarcoFalke replied:
Does providing a dtor make the compiler generated ctors go away?

hebasto replied:
> Does providing a dtor make the compiler generated ctors go away?

No, providing a destructor does not make the compiler generated constructors go away.
But the user-defined copy constructor (with this PR) makes the compiler generated default constructor go away.

My GCC fails to compile this PR without the default constructor.
It needs `FrozenCleanupCheck() {}` or `FrozenCleanupCheck() = default;`

MarcoFalke replied:
Idk. It passed for me after I removed all ctors

hebasto replied:
> ... I removed all ctors

That is the point ;)

Did Clang 10 fire any warning with passed `-Wdeprecated` ?

MarcoFalke replied:
I don't have clang 10, and it seems to be in development. I don't think we fix compiler warnings of compilers that are in alpha stage.

hebasto replied:
> I don't have clang 10, and it seems to be in development.

macOS 10.14 Mojave, XCode 10.0:
```
$ clang --version | grep version
Apple LLVM version 10.0.1 (clang-1001.0.46.4)
```

fanquake replied:
> Apple LLVM version 10.0.1 (clang-1001.0.46.4)

This version of Apple Clang is likely based off of ~ upstream Clang 6.0.1.

hebasto replied:
> Did Clang 10 fire any warning with passed `-Wdeprecated` ?

So, I mean a compiler on macOS 10.14+ ;)

MarcoFalke replied:
Well, thank god I don't have macOS, which is horribly broken in every thinkable way

hebasto commented:
@practicalswift Would you mind reviewing and testing this PR?
